### PR TITLE
Unicode block support

### PR DESCRIPTION
**Pull request by watzon**
This PR adds support for determining the UTF-8 code block of a unicode codepoint. If not appropriate here I can always make a library, but seeing as it's in the realm of UTF-8 support I figured I'd give it a go here first.

### CODE REVIEW

File: lib/std/unicode.zig [hunk 1]
```diff
@@ -741,3 +746,631 @@ test "utf8ToUtf16LeStringLiteral" {
...
+        0x2A700...0x2B73F => .CJKUnifiedIdeographsExtensionC,
+        0x2B740...0x2B81F => .CJKUnifiedIdeographsExtensionD,
+        0x2B820...0x2CEAF => .CJKUnifiedIdeographsExtensionE,
+        0x2CEB0...0x2EBEF => .CJKUnifiedIdeographsExtensionF,
+        0x2F800...0x2FA1F => .CJKCompatibilityIdeographsSupplement,
+        0xE0000...0xE007F => .Tags,
+        0xE0100...0xE01EF => .VariationSelectorsSupplement,
+        0xF0000...0xFFFFF => .SupplementaryPrivateUseAreaA,
+        0x100000...0x10FFFF => .SupplementaryPrivateUseAreaB,
+        else => error.InvalidUtf8
```
Comment (daurnimator) [inline lib/std/unicode.zig:1355]: There are still unassigned blocks; they won't be invalid in future.

Also the error is no utf8 specific, better to use `error.InvalidCodepoint` or something.

Comment (watzon) [reply lib/std/unicode.zig:1355]: Yeah that probably would be a better error

**Comment by watzon** [comment]
Knowing blocks is mainly useful in terms of parsing and filtering. For instance, filtering and validating emojis can be a pain, but with a list of all unicode blocks it becomes a bit easier. An `is_emoji` function could look like this:

```zig
pub fn is_emoji(c: u21) !bool {
    var block = try utf8BlockForCodepoint(c);
    return switch(block) {
        .MahjongTiles, .DominoTiles, PlayingCards, .EnclosedAlphanumericSupplement, .EnclosedIdeographicSupplement, .MiscellaneousSymbolsAndPictographs, .Emoticons, .OrnamentalDingbats, .TransportAndMapSymbols, .AlchemicalSymbols, .GeometricShapesExtended, .SupplementalArrowsC, .SupplementalSymbolsAndPictographs, .ChessSymbols, .SymbolsAndPictographsExtendedA => true,
        else => false
    }
}
```

Granted it's a bit long, but it would also allow you to filter out emojis by category. And this can be done with other unicode blocks such as Hiragana and Katakana.

**Comment by daurnimator** [comment]
> For instance, filtering and validating emojis can be a pain, but with a list of all unicode blocks it becomes a bit easier.

I don't think filtering by block is ever the correct thing to do.
Not all emojis are in that block (see https://www.unicode.org/Public/UCD/latest/ucd/emoji/emoji-data.txt); and there are other things in that block.

**Comment by watzon** [comment]
True, just an example. No method is super perfect unless you're willing to store that entire table and keep it up to date.

**Comment by data-man** [comment]
The names `ut8Block*` are slightly incorrect.
https://www.unicode.org/faq/blocks_ranges.html
> Blocks in the Unicode Standard are named ranges of **code points**.

---

Good Unicode support is very complex task.
BTW I'm working on Unicode proposal.
Some ideas:
Instead of hard coded ranges to introduce CodepointSet.
It's very useful, especially for regex engines and parsers.
E.g.:
```zig
const myLetters = unicode.Latin.add(Greek);
std.debug.assert(myLetters.has(0x0391));

const anyLetter = unicode.Lu.add(Ll);
std.debug.assert(myLetters.has(0x0391));

const l = CodepointSet.init(.{'a', 'A'});
std.debug.assert(anyLetter.has(l));
```

For future ideas: there is awesome [ucd-generate](https://github.com/BurntSushi/ucd-generate) for Rust. :)

**Comment by watzon** [comment]
@data-man that is an interesting idea. I look forward to seeing the proposal.

**Comment by data-man** [comment]
@watzon 
Also, why are you using Unicode 12 (or 12.1)?
UCD 13.0 has 308 blocks.

**Comment by data-man** [comment]
> that is an interesting idea.

In languages I used in the past, only D [has](https://dlang.org/phobos/std_uni.html#CodepointSet) CodepointSet in the standard library.

**Comment by watzon** [comment]
Hmm that's just the data set I found. If this is actually going to get merged I can update it, but I like your idea better.

**Comment by andrewrk** [comment]
This looks like it adds a dependency on Unicode data, which needs to go through the approval process before having a PR accepted for it. So far zig std lib does not have a dependency on any of the non-static components of unicode data (e.g. stuff that can change in future unicode updates). There have been several proposals rejected to add a dependency of the language on this.

**Pull request closed without merging**